core/src/Streamly/Internal/Data/Ring.hs

```diff
@@ -399,28 +361,25 @@ cast arr =
 -- the ring buffer. This is unsafe because the ringHead Ptr is not checked to
 -- be in range.
 {-# INLINE unsafeEqArrayN #-}
-unsafeEqArrayN :: Ring a -> Ptr a -> A.Array a -> Int -> Bool
+unsafeEqArrayN :: forall a. Unbox a => Ring a -> Int -> A.Array a -> Int -> Bool
 unsafeEqArrayN Ring{..} rh A.Array{..} nBytes
     | nBytes < 0 = error "unsafeEqArrayN: n should be >= 0"
     | nBytes == 0 = True
-    | otherwise = unsafeInlineIO $ check (castPtr rh) 0
+    | otherwise = unsafeInlineIO $ check (rh * SIZE_OF(a)) 0
 
     where
 
-    w8Contents = arrContents
-
     check p i = do
-        (relem :: Word8) <- peek p
-        aelem <- peekAt i w8Contents
+        (relem :: Word8) <- peekAt p ringContents
+        aelem <- peekAt i arrContents
         if relem == aelem
-        then go (p `plusPtr` 1) (i + 1)
+        then go (p + 1) (i + 1)
```

Comment:
No, this is correct. We are checking byte by byte. Not element by element. 
This is exactly like the previous code.
I did not want to introduce any semantic changes.
There must have been a reason we did not check this element by element in the previous code.
Possibly because deserialization of the entire element was expensive maybe?

Reply:
I missed the explicit typing to Word8. I thought we are using Unbox because of peekAt. We can remove the byte byte comparison and use memcmp instead. Can you create an issue for this?

Reply:
https://github.com/composewell/streamly/issues/2764